File: SGUEES-SPA/src/app/pages/Academics/aca-prospecto/aca-prospecto-contacto/aca-prospecto-contacto.service.ts
import { Observable } from 'rxjs';
import { Injectable } from '@angular/core';
import { IParam } from 'src/app/FxAPI/IParam';
import { IResult } from 'src/app/FxAPI/IResult';

import { AcaProspectoContactoRepository } from './aca-prospecto-contacto.repository';

// Qué hace: tipos de contacto que maneja la pantalla.
// Cómo lo hace: la tabla no usa catálogo, expresa el tipo con los bits ES_TELEFONO / ES_CORREO;
//               el formulario los presenta como un solo campo TIPO y el payload los traduce.
export const TIPO_TELEFONO = 'TELEFONO';
export const TIPO_CORREO = 'CORREO';

@Injectable({
    providedIn: 'root',
})
export class AcaProspectoContactoService {
    constructor(private repo: AcaProspectoContactoRepository) {}

    // Qué hace: correos y teléfonos del prospecto (grilla "Contacto").
    getContactosPorProspecto(CORR_PROSPECTO: number): Observable<IResult> {
        let xWhere: IParam[] = [{ Parameter: 'CORR_PROSPECTO', Value: CORR_PROSPECTO }];
        return this.repo.getAll(xWhere);
    }

    // Qué hace: modelo vacío para agregar un contacto.
    getModeloNuevo(persona: any): any {
        return this.normalizar({
            CORR_PROSPECTO_CONTACTO: 0,
            CORR_PROSPECTO_PERSONA: persona?.CORR_PROSPECTO_PERSONA ?? 0,
            CORR_PROSPECTO: persona?.CORR_PROSPECTO ?? 0,
            CONTACTO: '',
            ES_TELEFONO: true,
            ES_CORREO: false,
            ES_PRINCIPAL: false,
            ES_TRABAJO: false,
        });
    }

    // Qué hace: mantiene coherentes el campo TIPO (formulario), los bits (tabla) y el texto (grilla),
    //           y reparte un teléfono guardado en código de país + número.
    // Cómo lo hace: al cargar de la base TIPO no existe y se deduce de los bits; después manda TIPO,
    //               que es lo que el usuario elige, y los bits y el texto se derivan de él. El teléfono
    //               se guarda como en el registro ("+503 76150644"): si empieza con "+código " se
    //               separa; si es un teléfono viejo sin código, el código queda vacío y se pedirá.
    normalizar(model: any): any {
        if (model.TIPO !== TIPO_CORREO && model.TIPO !== TIPO_TELEFONO) {
            model.TIPO = model.ES_CORREO === true ? TIPO_CORREO : TIPO_TELEFONO;
        }
        model.ES_CORREO = model.TIPO === TIPO_CORREO;
        model.ES_TELEFONO = !model.ES_CORREO;
        model.TIPO_CONTACTO = model.ES_CORREO ? 'Correo' : 'Teléfono';

        if (model.ES_TELEFONO) {
            if (model.CODIGO_PAIS === undefined) {
                this.separarTelefono(model, model.CONTACTO);
            }
            // CONTACTO es lo que se guarda y lo que muestra la grilla: se rearma desde las dos partes.
            model.CONTACTO = this.getTelefonoCompleto(model);
        }
        return model;
    }

    // Qué hace: cambia el tipo del contacto sin perder lo escrito.
    // Cómo lo hace: al pasar a correo, el texto del número se ofrece como correo; al pasar a teléfono,
    //               el texto se separa en código y número (si no trae código, queda para elegirlo).
    cambiarTipo(model: any, tipo: string): any {
        if (tipo === TIPO_CORREO) {
            model.TIPO = TIPO_CORREO;
            model.CONTACTO = ((model.NUMERO ?? '').trim() || (model.CONTACTO ?? '').trim());
        } else {
            model.TIPO = TIPO_TELEFONO;
            this.separarTelefono(model, model.CONTACTO);
        }
        return this.normalizar(model);
    }

    // Qué hace: arma el teléfono completo con el formato del registro ("+503 76150644").
    getTelefonoCompleto(model: any): string {
        const numero = (model?.NUMERO ?? '').trim();
        const codigo = (model?.CODIGO_PAIS ?? '').trim();
        return codigo ? `${codigo} ${numero}`.trim() : numero;
    }

    // Qué hace: separa un teléfono guardado en código de país y número.
    private separarTelefono(model: any, valor: string): void {
        const texto = (valor ?? '').trim();
        const partes = /^(\+[\d+]+)\s+(.*)$/.exec(texto);
        model.CODIGO_PAIS = partes ? partes[1] : null;
        model.NUMERO = partes ? partes[2].trim() : texto;
    }

    // Qué hace: lista de códigos de país para el combo, a partir del SP del registro.
    // Cómo lo hace: el valor es el primer token del nombre ("+503"), que es lo que el registro guarda
    //               (no CatalogCode, que en ese catálogo no siempre coincide con la marcación). Dos
    //               países con el mismo código (+1 Canadá / Estados Unidos) se muestran juntos, porque
    //               un teléfono guardado no puede distinguirlos.
    getCodigosPais(catalogo: any[]): { CODIGO: string; NOMBRE: string }[] {
        const porCodigo = new Map<string, string[]>();
        for (const item of catalogo ?? []) {
            const nombre: string = (item.CatalogName ?? '').trim();
            const codigo = nombre.split(' ')[0];
            if (!codigo.startsWith('+')) {
                continue;
            }
            const pais = nombre.substring(codigo.length).trim();
            porCodigo.set(codigo, [...(porCodigo.get(codigo) ?? []), pais]);
        }
        return Array.from(porCodigo.entries())
            .map(([codigo, paises]) => ({ CODIGO: codigo, NOMBRE: `${codigo} ${paises.join(' / ')}` }))
            .sort((a, b) => a.NOMBRE.substring(a.CODIGO.length).localeCompare(b.NOMBRE.substring(b.CODIGO.length), 'es'));
    }

    insert(model: any): Observable<IResult> {
        return this.repo.create(this.getPayload(model));
    }

    update(model: any): Observable<IResult> {
        let xWhere: IParam[] = [{ Parameter: 'CORR_PROSPECTO_CONTACTO', Value: model.CORR_PROSPECTO_CONTACTO }];
        return this.repo.update(this.getPayload(model), xWhere);
    }

    delete(model: any): Observable<IResult> {
        let xWhere: IParam[] = [{ Parameter: 'CORR_PROSPECTO_CONTACTO', Value: model.CORR_PROSPECTO_CONTACTO }];
        return this.repo.delete(xWhere);
    }

    getColumns(): any {
        return [
            { dataField: 'TIPO_CONTACTO', caption: 'Tipo', width: 120 },
            { dataField: 'CONTACTO', caption: 'Contacto', width: 300 },
            {
                dataField: 'ES_PRINCIPAL',
                caption: 'Principal',
                width: 110,
                dataType: 'boolean',
                customizeText: (e: any) => (e.value ? 'Sí' : 'No'),
            },
            {
                dataField: 'ES_TRABAJO',
                caption: 'Trabajo',
                width: 110,
                dataType: 'boolean',
                customizeText: (e: any) => (e.value ? 'Sí' : 'No'),
            },
        ];
    }

    getSummary(): any {
        return {
            totalItems: [
                {
                    column: 'CORR_PROSPECTO_CONTACTO',
                    summaryType: 'count',
                    valueFormat: '#,##0',
                    displayFormat: 'Cant: {0}',
                },
            ],
        };
    }

    // Qué hace: formulario de un contacto (reemplaza a la grilla mientras se edita).
    // Cómo lo hace: el tipo siempre se puede corregir (un error de captura no debe dejar atrapado al
    //               usuario); el componente reacomoda al principal de cada tipo al aceptar. Un teléfono
    //               se captura como en el registro: código de país + número; un correo, en un solo campo.
    getItems(esPrincipalActual = false, esTelefono = true, codigosPais: any[] = []): any {
        return [
            {
                dataField: 'TIPO',
                label: { text: 'Tipo' },
                colSpan: 2,
                editorType: 'dxSelectBox',
                editorOptions: {
                    items: [
                        { valor: TIPO_TELEFONO, texto: 'Teléfono' },
                        { valor: TIPO_CORREO, texto: 'Correo' },
                    ],
                    valueExpr: 'valor',
                    displayExpr: 'texto',
                },
            },
            {
                dataField: 'CODIGO_PAIS',
                label: { text: 'Código de país' },
                colSpan: 2,
                editorType: 'dxSelectBox',
                editorOptions: {
                    items: codigosPais,
                    valueExpr: 'CODIGO',
                    displayExpr: 'NOMBRE',
                    searchEnabled: true,
                    searchExpr: 'NOMBRE',
                    showClearButton: true,
                    placeholder: 'Seleccionar',
                },
                visible: esTelefono,
                validationRules: esTelefono ? [{ type: 'required', message: 'Seleccione el código de país' }] : undefined,
            },
            {
                dataField: 'NUMERO',
                label: { text: 'Número' },
                colSpan: 2,
                editorOptions: { maxLength: 30 },
                visible: esTelefono,
                validationRules: esTelefono ? [{ type: 'required', message: 'Ingrese el número' }] : undefined,
            },
            {
                dataField: 'CONTACTO',
                label: { text: 'Correo' },
                colSpan: 4,
                editorOptions: { maxLength: 200 },
                visible: !esTelefono,
                validationRules: !esTelefono ? [{ type: 'required', message: 'Ingrese el correo' }] : undefined,
            },
            {
                dataField: 'ES_PRINCIPAL',
                label: { text: 'Principal de su tipo' },
                colSpan: 1,
                editorType: 'dxCheckBox',
                helpText: esPrincipalActual ? 'Es el principal: para quitarle la marca, marque otro' : undefined,
            },
            { dataField: 'ES_TRABAJO', label: { text: 'Del trabajo' }, colSpan: 1, editorType: 'dxCheckBox' },
        ];
    }

    // Qué hace: forma del dato según el tipo (misma regla que el API).
    // Cómo lo hace: correo = algo@algo.algo sin espacios; teléfono = código de país elegido y número
    //               con 7 a 12 dígitos (admite espacios, guiones y paréntesis), como en el registro.
    validar(model: any): string | null {
        if (model?.TIPO === TIPO_CORREO) {
            const valor = (model.CONTACTO ?? '').trim();
            if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(valor)) {
                return 'El correo no tiene un formato válido (ejemplo: nombre@dominio.com).';
            }
            return null;
        }

        if (!(model?.CODIGO_PAIS ?? '').trim()) {
            return 'Seleccione el código de país del teléfono.';
        }
        const numero = (model?.NUMERO ?? '').trim();
        const digitos = numero.replace(/[^0-9]/g, '').length;
        if (!/^[0-9\s()-]+$/.test(numero) || digitos < 7 || digitos > 12) {
            return 'El número solo admite dígitos, espacios, guiones y paréntesis, con 7 a 12 dígitos.';
        }
        return null;
    }

    // Qué hace: columnas de ACA_PROSPECTO_CONTACTO que viajan al API.
    // Cómo lo hace: TIPO se traduce a los bits; es público porque el componente lo usa además para
    //               saber si el contacto cambió.
    getPayload(model: any): any {
        const esCorreo = model.TIPO === TIPO_CORREO;
        return {
            CORR_PROSPECTO_CONTACTO: model.CORR_PROSPECTO_CONTACTO ?? 0,
            CORR_PROSPECTO_PERSONA: model.CORR_PROSPECTO_PERSONA,
            CONTACTO: esCorreo ? (model.CONTACTO ?? '').trim() : this.getTelefonoCompleto(model),
            ES_TELEFONO: !esCorreo,
            ES_CORREO: esCorreo,
            ES_PRINCIPAL: model.ES_PRINCIPAL === true,
            ES_TRABAJO: model.ES_TRABAJO === true,
        };
    }
}

File: SGUEES-API/sguees.api/Options/Academics/ACA_PROSPECTO_CONTACTO/ACA_PROSPECTO_CONTACTOService.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using sguees.Models;
using sguees.Repositories;

namespace sguees.Services
{
    public class ACA_PROSPECTO_CONTACTOService : IACA_PROSPECTO_CONTACTOService
    {
        private readonly IACA_PROSPECTO_CONTACTORepository _repo;

        public ACA_PROSPECTO_CONTACTOService(IACA_PROSPECTO_CONTACTORepository repo)
        {
            _repo = repo;
        }

        // Qué hace: correos y teléfonos del prospecto (pestaña Información personal).
        // Cómo lo hace: exige CORR_PROSPECTO porque eFramework omite del WHERE los enteros en 0
        //               y, sin él, devolvería los datos de todos los prospectos.
        public async Task<CResult> GetAllAsync(ACA_PROSPECTO_CONTACTOParam xWhere)
        {
            if (xWhere.CORR_PROSPECTO <= 0)
                return new CResult() { Data = null, Result = false, ErrorCode = -1, ErrorMessage = "Debe indicar el prospecto" };

            var p = new List<CParameter>
            {
                new CParameter() {ParameterName="CORR_EMPRESA",Value=xWhere.CORR_EMPRESA,DbType=System.Data.DbType.Int32},
                new CParameter() {ParameterName="CORR_PROSPECTO",Value=xWhere.CORR_PROSPECTO,DbType=System.Data.DbType.Int32},
            };

            return await _repo.GetAllAsync(p);
        }

        // Qué hace: un registro por su llave (CORR_PROSPECTO_CONTACTO); sin llave, el primero del prospecto.
        // Cómo lo hace: exige CORR_PROSPECTO por la misma razón que GetAllAsync;
        //               la llave es opcional porque eFramework omite los enteros en 0.
        public async Task<CResult> GetAsync(ACA_PROSPECTO_CONTACTOParam xWhere)
        {
            if (xWhere.CORR_PROSPECTO <= 0)
                return new CResult() { Data = null, Result = false, ErrorCode = -1, ErrorMessage = "Debe indicar el prospecto" };

            var p = new List<CParameter>
            {
                new CParameter() {ParameterName="CORR_EMPRESA",Value=xWhere.CORR_EMPRESA,DbType=System.Data.DbType.Int32},
                new CParameter() {ParameterName="CORR_PROSPECTO",Value=xWhere.CORR_PROSPECTO,DbType=System.Data.DbType.Int32},
                new CParameter() {ParameterName="CORR_PROSPECTO_CONTACTO",Value=xWhere.CORR_PROSPECTO_CONTACTO,DbType=System.Data.DbType.Int32},
            };

            return await _repo.GetAsync(p);
        }

        // Qué hace: códigos telefónicos de país, la misma lista del registro del portal.
        public async Task<CResult> GetCODIGO_PAIS_ACA_PROSPECTOAsync()
        {
            return await _repo.GetCODIGO_PAISAsync();
        }

        // Qué hace: valida y agrega un contacto del prospecto.
        public async Task<CResult> CreateAsync(ACA_PROSPECTO_CONTACTOTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            if (Data == null || Data.CORR_PROSPECTO_PERSONA <= 0)
                return ErrorValidacion("Debe indicar la persona del prospecto.");

            var validacion = Validar(Data);
            if (validacion != null) return validacion;

            return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        public async Task<CResult> UpdateAsync(ACA_PROSPECTO_CONTACTOTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            if (Data == null || Data.CORR_PROSPECTO_CONTACTO <= 0)
                return ErrorValidacion("Debe indicar el contacto a modificar.");

            var validacion = Validar(Data);
            if (validacion != null) return validacion;

            return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        public async Task<CResult> DeleteAsync(ACA_PROSPECTO_CONTACTOTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            if (Data == null || Data.CORR_PROSPECTO_CONTACTO <= 0)
                return ErrorValidacion("Debe indicar el contacto a eliminar.");

            return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: reglas comunes de alta y modificación del contacto.
        // Cómo lo hace: el dato es obligatorio (máximo 200) y el tipo es exactamente uno: teléfono o
        //               correo (la tabla lo expresa con bits, no con catálogo). Un correo debe tener
        //               forma de correo. Quién es el principal de cada tipo lo resuelve el repositorio.
        private static CResult Validar(ACA_PROSPECTO_CONTACTOTable Data)
        {
            Data.CONTACTO = string.IsNullOrWhiteSpace(Data.CONTACTO) ? null : Data.CONTACTO.Trim();

            if (Data.CONTACTO == null)
                return ErrorValidacion("Ingrese el teléfono o el correo.");
            if (Data.CONTACTO.Length > 200)
                return ErrorValidacion("El contacto no puede superar 200 caracteres.");
            if (Data.ES_TELEFONO == Data.ES_CORREO)
                return ErrorValidacion("Indique si el contacto es un teléfono o un correo.");
            if (Data.ES_CORREO && !EsCorreo(Data.CONTACTO))
                return ErrorValidacion("El correo no tiene un formato válido (ejemplo: nombre@dominio.com).");
            if (Data.ES_TELEFONO && !EsTelefono(Data.CONTACTO))
                return ErrorValidacion("El teléfono debe llevar código de país y de 7 a 12 dígitos (ejemplo: +503 76150644).");

            return null;
        }

        // Qué hace: forma de correo: algo@algo.algo, sin espacios ni segunda arroba.
        private static bool EsCorreo(string valor) =>
            System.Text.RegularExpressions.Regex.IsMatch(valor, @"^[^\s@]+@[^\s@]+\.[^\s@]+$");

        // Qué hace: forma de teléfono igual a la del registro del portal: "+503 76150644", es decir,
        //           código de país (primer token del catálogo, puede ser "+1+767"), un espacio y el
        //           número con 7 a 12 dígitos (admite espacios, guiones y paréntesis).
        private static bool EsTelefono(string valor)
        {
            var espacio = valor.IndexOf(' ');
            if (espacio <= 0) return false;

            var codigo = valor.Substring(0, espacio);
            var numero = valor.Substring(espacio + 1).Trim();
            if (!System.Text.RegularExpressions.Regex.IsMatch(codigo, @"^\+\d{1,4}(\+\d{1,4})?$")) return false;

            var digitos = 0;
            foreach (var c in numero)
            {
                if (char.IsDigit(c)) { digitos++; continue; }
                if (c == ' ' || c == '-' || c == '(' || c == ')') continue;
                return false;
            }
            return digitos >= 7 && digitos <= 12;
        }

        private static CResult ErrorValidacion(string mensaje)
        {
            return new CResult() { Data = null, Result = false, CodeHelper = 0, ErrorCode = -1, ErrorMessage = mensaje, ErrorSource = "[ACA_PROSPECTO_CONTACTOService]", RowsAffected = 0 };
        }
    }
}
